### PR TITLE
Add table64 lowering pass

### PR DESCRIPTION
**Pull request by sbc100**
This needs to land before I can update llvm: https://github.com/llvm/llvm-project/pull/92042

### CODE REVIEW

File: src/passes/Table64Lowering.cpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Lowers a module with a 64-bit table to one with a 32-bit table.
+//
+// This pass can be delete once table64 is implemented in the Wasm engines:
```
Comment (kripken) [inline src/passes/Table64Lowering.cpp:20]:
```suggestion
// This pass can be deleted once table64 is implemented in Wasm engines:
```

File: src/passes/Table64Lowering.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  void wrapAddress64(Expression*& ptr, Name tableName) {
+    if (ptr->type == Type::unreachable) {
+      return;
+    }
+    auto& module = *getModule();
+    auto table = module.getTable(tableName);
+    if (table->is64()) {
+      assert(ptr->type == Type::i64);
+      Builder builder(module);
+      ptr = builder.makeUnary(UnaryOp::WrapInt64, ptr);
```
Comment (kripken) [inline src/passes/Table64Lowering.cpp:46]:
```suggestion
      ptr = Builder(module).makeUnary(UnaryOp::WrapInt64, ptr);
```

File: src/passes/Table64Lowering.cpp [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+static Name TABLE_BASE32("__table_base32");
+
+struct Table64Lowering : public WalkerPass<PostWalker<Table64Lowering>> {
+
+  void wrapAddress64(Expression*& ptr, Name tableName) {
+    if (ptr->type == Type::unreachable) {
+      return;
+    }
+    auto& module = *getModule();
+    auto table = module.getTable(tableName);
```
Comment (kripken) [inline src/passes/Table64Lowering.cpp:42]:
```suggestion
    auto* table = module.getTable(tableName);
```

File: src/passes/Table64Lowering.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+      ptr = builder.makeUnary(UnaryOp::WrapInt64, ptr);
+    }
+  }
+
+  void extendAddress64(Expression*& ptr, Name tableName) {
+    if (ptr->type == Type::unreachable) {
+      return;
+    }
+    auto& module = *getModule();
+    auto table = module.getTable(tableName);
```
Comment (kripken) [inline src/passes/Table64Lowering.cpp:55]:
```suggestion
    auto* table = module.getTable(tableName);
```

File: src/passes/Table64Lowering.cpp [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    if (table->is64()) {
+      assert(ptr->type == Type::i64);
+      ptr->type = Type::i32;
+      Builder builder(module);
+      ptr = builder.makeUnary(UnaryOp::ExtendUInt32, ptr);
```
Comment (kripken) [inline src/passes/Table64Lowering.cpp:60]:
```suggestion
      ptr = Builder(module).makeUnary(UnaryOp::ExtendUInt32, ptr);
```

File: src/passes/Table64Lowering.cpp [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    }
+  }
+
+  void visitTableSize(TableSize* curr) {
+    auto& module = *getModule();
+    auto table = module.getTable(curr->table);
+    if (table->is64()) {
+      auto size = static_cast<Expression*>(curr);
```
Comment (kripken) [inline src/passes/Table64Lowering.cpp:68]:
```suggestion
      auto* size = static_cast<Expression*>(curr);
```

File: src/passes/Table64Lowering.cpp [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+      assert(ptr->type == Type::i64);
+      ptr->type = Type::i32;
+      Builder builder(module);
+      ptr = builder.makeUnary(UnaryOp::ExtendUInt32, ptr);
+    }
+  }
+
+  void visitTableSize(TableSize* curr) {
+    auto& module = *getModule();
+    auto table = module.getTable(curr->table);
```
Comment (kripken) [inline src/passes/Table64Lowering.cpp:66]:
```suggestion
    auto* table = module.getTable(curr->table);
```

File: src/passes/Table64Lowering.cpp (hunk 6 above)
Comment (kripken) [reply src/passes/Table64Lowering.cpp:68]: Though we more commonly write
```suggestion
      Expression* size = curr;
```
But maybe static_cast is better?

File: src/passes/Table64Lowering.cpp [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    if (table->is64()) {
+      auto size = static_cast<Expression*>(curr);
+      extendAddress64(size, curr->table);
+      replaceCurrent(size);
+    }
+  }
+
+  void visitTableGrow(TableGrow* curr) {
+    auto& module = *getModule();
+    auto table = module.getTable(curr->table);
```
Comment (kripken) [inline src/passes/Table64Lowering.cpp:76]:
```suggestion
    auto* table = module.getTable(curr->table);
```

File: test/lit/passes/table64-lowering-features.wast [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+(module
+  ;; CHECK:      (type $0 (func))
+
+  ;; CHECK:      (type $1 (func (result i64)))
+
+  ;; CHECK:      (table $t 10 100 funcref)
+  (table $t i64 10 100 funcref)
+
+  (elem (table $t) (i64.const 0) funcref (ref.null func))
```
Comment (kripken) [inline test/lit/passes/table64-lowering-features.wast:13]: Channelling @tlively , if you give this `elem` a name then the CHECK will appear right on top of it :smile: I always forget that myself.
```suggestion
  (elem $elem (table $t) (i64.const 0) funcref (ref.null func))
```

File: src/passes/Table64Lowering.cpp (hunk 3 above)
Comment (sbc100) [reply src/passes/Table64Lowering.cpp:42]: This was copied over the the memory lowering pass.. updated there too.

File: src/passes/Table64Lowering.cpp (hunk 2 above)
Comment (sbc100) [reply src/passes/Table64Lowering.cpp:46]: Ditto.

**Comment by sbc100** [comment]
Updated.  And with feedback in a separate commit!

**Review by kripken — APPROVED** [review]

**Comment by sbc100** [comment]
Any idea whats up with the tsan failure.  It doesn't look related:

```
n file included from /home/runner/work/binaryen/binaryen/src/passes/StringLowering.cpp:33:
124
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/algorithm:61:
125
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_algo.h:61:
126
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_tempbuf.h:263:8: error: 'get_temporary_buffer<std::unique_ptr<wasm::Global>>' is deprecated [-Werror,-Wdeprecated-declarations]
127
  263 |                 std::get_temporary_buffer<value_type>(_M_original_len));
128
      |                      ^
129
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_algo.h:4996:15: note: in instantiation of member function 'std::_Temporary_buffer<__gnu_cxx::__normal_iterator<std::unique_ptr<wasm::Global> *, std::vector<std::unique_ptr<wasm::Global>>>, std::unique_ptr<wasm::Global>>::_Temporary_buffer' requested here
130
 4996 |       _TmpBuf __buf(__first, (__last - __first + 1) / 2);
131
      |               ^
132
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_algo.h:5070:23: note: in instantiation of function template specialization 'std::__stable_sort<__gnu_cxx::__normal_iterator<std::unique_ptr<wasm::Global> *, std::vector<std::unique_ptr<wasm::Global>>>, __gnu_cxx::__ops::_Iter_comp_iter<(lambda at /home/runner/work/binaryen/binaryen/src/passes/StringLowering.cpp:174:7)>>' requested here
133
 5070 |       _GLIBCXX_STD_A::__stable_sort(__first, __last,
134
      |                       ^
135
/home/runner/work/binaryen/binaryen/src/passes/StringLowering.cpp:171:10: note: in instantiation of function template specialization 'std::stable_sort<__gnu_cxx::__normal_iterator<std::unique_ptr<wasm::Global> *, std::vector<std::unique_ptr<wasm::Global>>>, (lambda at /home/runner/work/binaryen/binaryen/src/passes/StringLowering.cpp:174:7)>' requested here
136
  171 |     std::stable_sort(
137
      |          ^
138
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_tempbuf.h:99:5: note: 'get_temporary_buffer<std::unique_ptr<wasm::Global>>' has been explicitly marked deprecated here
139
   99 |     _GLIBCXX17_DEPRECATED
140
      |     ^
141
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/x86_64-linux-gnu/c++/12/bits/c++config.h:119:34: note: expanded from macro '_GLIBCXX17_DEPRECATED'
142
  119 | # define _GLIBCXX17_DEPRECATED [[__deprecated__]]
143
      |                                  ^
144
1 error generated.
```

**Comment by sbc100** [comment]
The tsan and asan failure apprear to be issues that also exist in main.. so I'm tempted land this despite them.

**Comment by tlively** [comment]
Yeah, looks like something on the runners changed. I think landing this is fine.

**Pull request merged**